stress: failed test in cockroach/sql/distsql/distsql.test: TestTableReader

Binary: cockroach/static-tests.tar.gz sha: https://github.com/cockroachdb/cockroach/commits/d9039b883cc2c9190448216f57ccfe4676d4e7e1

Stress build found a failed test:

```
=== RUN   TestTableReader
I160614 05:57:53.366186 storage/engine/rocksdb.go:137  opening in memory rocksdb instance
W160614 05:57:53.372117 gossip/gossip.go:897  not connected to cluster; use --join to specify a connected node
I160614 05:57:53.373241 server/node.go:401  store store=0:0 ([]=) not bootstrapped
W160614 05:57:53.373448 gossip/gossip.go:897  not connected to cluster; use --join to specify a connected node
I160614 05:57:53.375666 storage/replica_command.go:1531  range 1: new leader lease replica {1 1 1} 1970-01-01 00:00:00 +0000 UTC 407189h57m57.875247282s following replica {0 0 0} 1970-01-01 00:00:00 +0000 UTC 0 [physicalTime=2016-06-14 05:57:53.375569066 +0000 UTC]
I160614 05:57:53.376225 server/node.go:327  **** cluster {c6f6e07e-6f16-4f25-a41b-329da7c01a8c} has been created
I160614 05:57:53.376245 server/node.go:328  **** add additional nodes by specifying --join=127.0.0.1:38275
W160614 05:57:53.376256 gossip/gossip.go:897  not connected to cluster; use --join to specify a connected node
I160614 05:57:53.376923 server/node.go:414  initialized store store=1:1 ([]=): {Capacity:10365558784 Available:8197558272 RangeCount:0}
I160614 05:57:53.376943 server/node.go:302  node ID 1 initialized
I160614 05:57:53.376991 storage/stores.go:287  read 0 node addresses from persistent storage
I160614 05:57:53.377060 server/node.go:535  connecting to gossip network to verify cluster ID...
I160614 05:57:53.377782 server/node.go:556  node connected via gossip and verified as part of cluster {"c6f6e07e-6f16-4f25-a41b-329da7c01a8c"}
I160614 05:57:53.377806 server/node.go:355  [node=1] Started node with [[]=] engine(s) and attributes []
I160614 05:57:53.377827 server/server.go:417  starting https server at 127.0.0.1:40283
I160614 05:57:53.377836 server/server.go:418  starting grpc/postgres server at 127.0.0.1:38275
I160614 05:57:53.378162 storage/split_queue.go:96  splitting range=1 [/Min-/Max) at keys [/Table/11/0 /Table/12/0 /Table/13/0 /Table/14/0]
I160614 05:57:53.406799 server/updates.go:156  No previous updates check time.
I160614 05:57:53.407526 storage/replica_command.go:2079  initiating a split of range=1 [/Min-/Max) at key /Table/11
I160614 05:57:53.426170 storage/replica_command.go:1531  range 2: new leader lease replica {1 1 1} 1970-01-01 00:00:00 +0000 UTC 407189h57m57.976563251s following replica {0 0 0} 1970-01-01 00:00:00 +0000 UTC 0 [physicalTime=2016-06-14 05:57:53.426108341 +0000 UTC]
I160614 05:57:53.426317 storage/replica_command.go:2079  initiating a split of range=2 [/Table/11-/Max) at key /Table/12
I160614 05:57:53.433240 storage/replica_command.go:1531  range 3: new leader lease replica {1 1 1} 1970-01-01 00:00:00 +0000 UTC 407189h57m57.982922771s following replica {0 0 0} 1970-01-01 00:00:00 +0000 UTC 0 [physicalTime=2016-06-14 05:57:53.433201941 +0000 UTC]
I160614 05:57:53.433400 storage/replica_command.go:2079  initiating a split of range=3 [/Table/12-/Max) at key /Table/13
I160614 05:57:53.461374 storage/replica_command.go:1531  range 4: new leader lease replica {1 1 1} 1970-01-01 00:00:00 +0000 UTC 407189h57m58.011089201s following replica {0 0 0} 1970-01-01 00:00:00 +0000 UTC 0 [physicalTime=2016-06-14 05:57:53.461335722 +0000 UTC]
I160614 05:57:53.461568 storage/replica_command.go:2079  initiating a split of range=4 [/Table/13-/Max) at key /Table/14
I160614 05:57:53.537285 storage/replica_command.go:1531  range 5: new leader lease replica {1 1 1} 1970-01-01 00:00:00 +0000 UTC 407189h57m58.087057453s following replica {0 0 0} 1970-01-01 00:00:00 +0000 UTC 0 [physicalTime=2016-06-14 05:57:53.537240115 +0000 UTC]
I160614 05:57:53.537428 storage/split_queue.go:96  splitting range=5 [/Table/14-/Max) at keys [/Table/50/0]
I160614 05:57:53.548020 storage/replica_command.go:2079  initiating a split of range=5 [/Table/14-/Max) at key /Table/50
I160614 05:57:53.587458 stopper.go:352  draining; tasks left:
1      storage/queue.go:386
1      server/node.go:803
1      kv/txn_coord_sender.go:870
I160614 05:57:53.587519 stopper.go:352  draining; tasks left:
1      storage/queue.go:386
1      server/node.go:803
E160614 05:57:53.598208 client/txn.go:370  failure aborting transaction: writing transaction timed out or ran on multiple coordinators; abort caused by: node unavailable; try another peer
E160614 05:57:53.598462 storage/queue.go:449  [split] range=5 [/Table/14-/Max): error: storage/split_queue.go:99: unable to split range=5 [/Table/14-/Max) at key "/Table/50/0": storage/replica_command.go:2132: split at key /Table/50 failed: node unavailable; try another peer
I160614 05:57:54.589724 /go/src/google.golang.org/grpc/clientconn.go:499  grpc: Conn.resetTransport failed to create client transport: connection error: desc = "transport: dial tcp 127.0.0.1:38275: getsockopt: connection refused"; Reconnecting to "127.0.0.1:38275"
--- FAIL: TestTableReader (5.27s)
	leaktest.go:86: Leaked goroutine: goroutine 336 [IO wait]:
		net.runtime_pollWait(0x7f3b9a701608, 0x72, 0xc8207ef000)
			/usr/local/go/src/runtime/netpoll.go:160 +0x60
		net.(*pollDesc).Wait(0xc8207fe290, 0x72, 0x0, 0x0)
			/usr/local/go/src/net/fd_poll_runtime.go:73 +0x3a
		net.(*pollDesc).WaitRead(0xc8207fe290, 0x0, 0x0)
			/usr/local/go/src/net/fd_poll_runtime.go:78 +0x36
		net.(*netFD).Read(0xc8207fe230, 0xc8207ef000, 0x400, 0x400, 0x0, 0x7f3b9a775050, 0xc82000e068)
			/usr/local/go/src/net/fd_unix.go:250 +0x23a
		net.(*conn).Read(0xc820578010, 0xc8207ef000, 0x400, 0x400, 0x0, 0x0, 0x0)
			/usr/local/go/src/net/net.go:172 +0xe4
		crypto/tls.(*block).readFromUntil(0xc8204fa120, 0x7f3b9a781db8, 0xc820578010, 0x5, 0x0, 0x0)
			/usr/local/go/src/crypto/tls/conn.go:460 +0xcc
		crypto/tls.(*Conn).readRecord(0xc82024c300, 0x1ce9217, 0x0, 0x0)
			/usr/local/go/src/crypto/tls/conn.go:562 +0x2d1
		crypto/tls.(*Conn).Read(0xc82024c300, 0xc820642000, 0x8000, 0x8000, 0x0, 0x0, 0x0)
			/usr/local/go/src/crypto/tls/conn.go:939 +0x167
		bufio.(*Reader).fill(0xc820423320)
			/usr/local/go/src/bufio/bufio.go:97 +0x1e9
		bufio.(*Reader).Read(0xc820423320, 0xc8202ecff8, 0x9, 0x9, 0x9, 0x0, 0x0)
			/usr/local/go/src/bufio/bufio.go:207 +0x260
		io.ReadAtLeast(0x7f3b9a6765e8, 0xc820423320, 0xc8202ecff8, 0x9, 0x9, 0x9, 0x0, 0x0, 0x0)
			/usr/local/go/src/io/io.go:297 +0xe6
		io.ReadFull(0x7f3b9a6765e8, 0xc820423320, 0xc8202ecff8, 0x9, 0x9, 0xcea8d0d93dfda001, 0x0, 0x0)
			/usr/local/go/src/io/io.go:315 +0x62
		golang.org/x/net/http2.readFrameHeader(0xc8202ecff8, 0x9, 0x9, 0x7f3b9a6765e8, 0xc820423320, 0x9a000000, 0xc800000000, 0x0, 0x0)
			/go/src/golang.org/x/net/http2/frame.go:237 +0xa5
		golang.org/x/net/http2.(*Framer).ReadFrame(0xc8202ecfc0, 0x0, 0x0, 0x0, 0x0)
			/go/src/golang.org/x/net/http2/frame.go:464 +0x106
		google.golang.org/grpc/transport.(*framer).readFrame(0xc82078ca80, 0x0, 0x0, 0x0, 0x0)
			/go/src/google.golang.org/grpc/transport/http_util.go:417 +0x3d
		google.golang.org/grpc/transport.(*http2Client).reader(0xc8204a63c0)
			/go/src/google.golang.org/grpc/transport/http2_client.go:795 +0x109
		created by google.golang.org/grpc/transport.newHTTP2Client
			/go/src/google.golang.org/grpc/transport/http2_client.go:174 +0xd21
	leaktest.go:86: Leaked goroutine: goroutine 337 [select]:
		google.golang.org/grpc/transport.(*http2Client).controller(0xc8204a63c0)
			/go/src/google.golang.org/grpc/transport/http2_client.go:873 +0x5da
		created by google.golang.org/grpc/transport.newHTTP2Client
			/go/src/google.golang.org/grpc/transport/http2_client.go:201 +0x15c2
	leaktest.go:86: Leaked goroutine: goroutine 367 [select]:
		google.golang.org/grpc.(*Conn).transportMonitor(0xc820052000)
			/go/src/google.golang.org/grpc/clientconn.go:547 +0x1d3
		google.golang.org/grpc.NewConn.func1(0xc820052000)
			/go/src/google.golang.org/grpc/clientconn.go:355 +0x1b5
		created by google.golang.org/grpc.NewConn
			/go/src/google.golang.org/grpc/clientconn.go:356 +0x4e3
FAIL


ERROR: exit status 1
```

Run Details:

```
74 runs so far, 0 failures, over 5s
159 runs so far, 0 failures, over 10s
241 runs so far, 0 failures, over 15s
326 runs so far, 0 failures, over 20s
407 runs so far, 0 failures, over 25s
488 runs so far, 0 failures, over 30s
570 runs so far, 0 failures, over 35s
650 runs so far, 0 failures, over 40s
731 runs so far, 0 failures, over 45s
813 runs so far, 0 failures, over 50s
895 runs so far, 0 failures, over 55s
979 runs so far, 0 failures, over 1m0s
1062 runs so far, 0 failures, over 1m5s
1142 runs so far, 0 failures, over 1m10s
1224 runs so far, 0 failures, over 1m15s
1306 runs so far, 0 failures, over 1m20s
1389 runs so far, 0 failures, over 1m25s
1471 runs so far, 0 failures, over 1m30s
1546 runs so far, 0 failures, over 1m35s
1629 runs so far, 0 failures, over 1m40s
1716 runs so far, 0 failures, over 1m45s
1797 runs so far, 0 failures, over 1m50s
1881 runs so far, 0 failures, over 1m55s
1963 runs so far, 0 failures, over 2m0s
2047 runs so far, 0 failures, over 2m5s
2127 runs so far, 0 failures, over 2m10s
2208 runs so far, 0 failures, over 2m15s
2291 runs so far, 0 failures, over 2m20s
2373 runs so far, 0 failures, over 2m25s
2450 runs so far, 0 failures, over 2m30s
2537 runs so far, 0 failures, over 2m35s
2618 runs so far, 0 failures, over 2m40s
2700 runs so far, 0 failures, over 2m45s
2781 runs so far, 0 failures, over 2m50s
2860 runs so far, 0 failures, over 2m55s
2941 runs so far, 0 failures, over 3m0s
3024 runs so far, 0 failures, over 3m5s
3105 runs so far, 0 failures, over 3m10s
3187 runs so far, 0 failures, over 3m15s
3268 runs so far, 0 failures, over 3m20s
3352 runs so far, 0 failures, over 3m25s
3433 runs so far, 0 failures, over 3m30s
3517 runs so far, 0 failures, over 3m35s
3603 runs so far, 0 failures, over 3m40s
3684 runs so far, 0 failures, over 3m45s
3769 runs so far, 0 failures, over 3m50s
3849 runs so far, 0 failures, over 3m55s
3935 runs so far, 0 failures, over 4m0s
4014 runs so far, 0 failures, over 4m5s
4094 runs so far, 0 failures, over 4m10s
4177 runs so far, 0 failures, over 4m15s
4258 runs so far, 0 failures, over 4m20s
4341 runs so far, 0 failures, over 4m25s
4426 runs so far, 0 failures, over 4m30s
4507 runs so far, 0 failures, over 4m35s
4589 runs so far, 0 failures, over 4m40s
4671 runs so far, 0 failures, over 4m45s
4755 runs so far, 0 failures, over 4m50s
4837 runs so far, 0 failures, over 4m55s
4921 runs so far, 0 failures, over 5m0s
5000 runs so far, 0 failures, over 5m5s
5082 runs so far, 0 failures, over 5m10s
5168 runs so far, 0 failures, over 5m15s
5246 runs so far, 0 failures, over 5m20s
5332 runs so far, 0 failures, over 5m25s
5412 runs so far, 0 failures, over 5m30s
5496 runs so far, 0 failures, over 5m35s
5578 runs so far, 0 failures, over 5m40s
5661 runs so far, 0 failures, over 5m45s
5747 runs so far, 0 failures, over 5m50s
5826 runs so far, 0 failures, over 5m55s
5907 runs so far, 0 failures, over 6m0s
5990 runs so far, 0 failures, over 6m5s
6073 runs so far, 0 failures, over 6m10s
6157 runs so far, 0 failures, over 6m15s
6239 runs so far, 0 failures, over 6m20s
6320 runs so far, 0 failures, over 6m25s
6403 runs so far, 0 failures, over 6m30s
6484 runs so far, 0 failures, over 6m35s
6567 runs so far, 0 failures, over 6m40s
6651 runs so far, 0 failures, over 6m45s
6733 runs so far, 0 failures, over 6m50s
6812 runs so far, 0 failures, over 6m55s
6892 runs so far, 0 failures, over 7m0s
6976 runs so far, 0 failures, over 7m5s
7059 runs so far, 0 failures, over 7m10s
7142 runs so far, 0 failures, over 7m15s
7226 runs so far, 0 failures, over 7m20s
7307 runs so far, 0 failures, over 7m25s
7391 runs so far, 0 failures, over 7m30s
7474 runs so far, 0 failures, over 7m35s
7553 runs so far, 0 failures, over 7m40s
7636 runs so far, 0 failures, over 7m45s
7718 runs so far, 0 failures, over 7m50s
7802 runs so far, 0 failures, over 7m55s
7882 runs so far, 0 failures, over 8m0s
7962 runs so far, 0 failures, over 8m5s
8044 runs so far, 0 failures, over 8m10s
8128 runs so far, 0 failures, over 8m15s
8211 runs so far, 0 failures, over 8m20s
8295 runs so far, 0 failures, over 8m25s
8375 runs so far, 0 failures, over 8m30s
8459 runs so far, 0 failures, over 8m35s
8538 runs so far, 0 failures, over 8m40s
8620 runs so far, 0 failures, over 8m45s
8704 runs so far, 0 failures, over 8m50s
8784 runs so far, 0 failures, over 8m55s
8800 runs completed, 1 failures, over 8m56s
FAIL
```
Please assign, take a look and update the issue accordingly.

--- a/rpc/context.go
+++ b/rpc/context.go
@@ -167,7 +167,7 @@ func (ctx *Context) GRPCDial(target string, opts ...grpc.DialOption) (*grpc.Clie
 		}
 		ctx.conns.cache[target] = conn
 
-		ctx.Stopper.RunTask(func() {
+		if !ctx.Stopper.RunTask(func() {
 			ctx.Stopper.RunWorker(func() {
 				if err := ctx.runHeartbeat(conn, target); err != nil && !grpcutil.IsClosedConnection(err) {
 					log.Error(err)
@@ -176,7 +176,9 @@ func (ctx *Context) GRPCDial(target string, opts ...grpc.DialOption) (*grpc.Clie
 				ctx.removeConn(target, conn)
 				ctx.conns.Unlock()
 			})
-		})
+		}) {
+			ctx.removeConn(target, conn)
+		}
 	}
 	return conn, err
 }

